[Common Lisp] Add new concept exercise: arithmetic (#1942)
exercism/v3: languages/common-lisp/exercises/concept/arithmetic/.docs/introduction.md

## Numerical Types

> In Common Lisp, like many languages, numbers come in a couple of types. This
> exercise focuses on getting to grips with two of the most basic number-types:

Reviewer: The `introduction.md` is usually written as just being about the Concept(s) involved, without any reference to the exercise. It is not wrong per se, but maybe do:

```suggestion
In Common Lisp, like many languages, numbers come in a couple of types. The two most basic number-types are:
```

## Numerical Types

In Common Lisp, like many languages, numbers come in a couple of types. This
exercise focuses on getting to grips with two of the most basic number-types:
- Integers: whole numbers without a decimal point (like `-6`, `0`, `25`, `1234`,
  etc.)
- Floating-Point Numbers: fractional or whole numbers including a decimal point (like
  `3.14`, `-1.7`, `99.99`, `2048.0`)

In general, if you are working with only whole numbers, you should prefer
integers as they don't suffer from the same loss of precision as floating-point
numbers do over many calculations.

## Arithmetic & Prefix Notation

Common Lisp uses the standard arithmetic operators for most operations (you can
find a [full list here](http://clhs.lisp.se/Body/12_aa.htm)), but is somewhat
unique in using a "prefix-notation" as opposed to the more familiar
"infix-notion". More visually:

```lisp
;; Infix-notation (non-lisp languages)
1 + 2 + 3 + 4 + 5 ; => 15
;; Prefix-notation (lisp languages)
(+ 1 2 3 4 5) ; => 15
```

While prefix notion turns some operations like `2 + 2` into the somewhat
unfamiliar `(+ 2 2)` form, it makes it much easier to operate on more than one

> number at a time.

Reviewer 2: I think a quick example here of 'more than one number' would be good, in case the reader cannot easily imaging `(+ 1 2 3 4 5 6 7 8 9)` as a thing one could do in a programming language.

Author: I showed that in the code snippet above under prefix notation. Do you think I should highlight that a bit more?

Reviewer 2: Hmm if it was already there then i just missed it. nevermind.

As a final quirk, the `-` and `/` operators have a special meanings when applied
to only one number:

```lisp
;; A single number passed to `-` is simply negated
(- 4)   ; => -4
(- -32) ; => 32
;; A single number passed to `/` returns the reciprocal
(/ 8)   ; => 1/8
(/ 0.1) ; => 10.0
```

## Comparing Numbers

Finally, you may find it useful to compare different numbers using functions
like `=`, `/=` (not equal to), and `>=` (greater than or equal to). When these
comparisons are true (as in `(= 1 1)`), they return `T` and when they aren't (as
in `(> 0 1)`), they return `NIL`. For a full list of comparison operators, see

> Figure 12-3 [here](http://clhs.lisp.se/Body/12_aa.htm).

Reviewer 2: I tend to use hyperspec links pointing to lispworks (becuase l1sp.org redirector does that). We should probably decide upon one Hyperspec instance to link to and standardize.

Author: I think I like the lispworks version better actually! I think I'll move to that to be in-line with your exercises!

Author: Sorry to reopen this, but I think I was a bit confused the first time around. I think that the basics exercise and all of mine so far do point to clhs.lisp.se which I might prefer if for no other reason than the links being shorter.

Is there some reason to prefer lispworks urls?

Reviewer 2: I use the Lispworks link 1) because that is where l1sp.org's redirector points (so i can quickly create links to things (e.g. http://l1sp.org/cl/expt ) and 2) they are the owners of the coypright for the HyperSpec. (http://www.lispworks.com/documentation/HyperSpec/Front/Help.htm#Legal)

Who is the owner/maintainer of lisp.se?

As for shorter links we could use l1sp.org links in some cases.

Other than length - what leads you to prefer lisp.se?

Author: Actually, do you find l1sp.org to be rather reliable? That seems like a good service to use as we can point people loads of places with it and the URLs are probably the most meaningful. If I want to direct people to the comparison operators, it doesn't get much simpler than http://l1sp.org/cl/=

I'd be happy with using l1sp.org links if you think that's reliable.

Reviewer 2: Yes, i find l1sp.org very reliable. I don't think i've found a case of it being down.

Now for this track we could use the l1sp links or the lispwork links. Because of the 'danger' of using link shorteners (the service may go away) I might lean toward using lispworks since i assume that company will be more stable than Xach (a single person living in Maine USA, supporting this in his spare time on their own money).

> ## Numerical Types

Reviewer: Nice introduction!

In Common Lisp, like many languages, numbers come in a couple of types. This
exercise focuses on getting to grips with two of the most basic number-types:

- Integers: whole numbers without a decimal point (like `-6`, `0`, `25`, `1234`,
  etc.)
- Floating-Point Numbers: fractional or whole numbers including a decimal point (like
  `3.14`, `-1.7`, `99.99`, `2048.0`)

In general, if you are working with only whole numbers, you should prefer
integers as they don't suffer from the same loss of precision as floating-point
numbers do over many calculations.

## Arithmetic & Prefix Notation

> Common Lisp uses the standard arithmetic operators for most operations (you can

Reviewer: The [docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd) state:

> Links should be used sparingly, if at all. While a link explaining a complex topic like recursion might be useful, for most concepts the links will provide more information than needed so explaining things concisely inline should be the aim.

Maybe remove the link to the arithmetic operators? It could be in the `hints.md` and/or `after.md` instead.

find a [full list here](http://clhs.lisp.se/Body/12_aa.htm)), but is somewhat
unique in using a "prefix-notation" as opposed to the more familiar
"infix-notion". More visually:

```lisp
;; Infix-notation (non-lisp languages)
1 + 2 + 3 + 4 + 5 ; => 15
;; Prefix-notation (lisp languages)
(+ 1 2 3 4 5) ; => 15
```

While prefix notion turns some operations like `2 + 2` into the somewhat
unfamiliar `(+ 2 2)` form, it makes it much easier to operate on more than one
number at a time.

As a final quirk, the `-` and `/` operators have a special meanings when applied
to only one number:

```lisp
;; A single number passed to `-` is simply negated
(- 4)   ; => -4
(- -32) ; => 32
;; A single number passed to `/` returns the reciprocal
(/ 8)   ; => 1/8
(/ 0.1) ; => 10.0
```

## Comparing Numbers

Finally, you may find it useful to compare different numbers using functions
like `=`, `/=` (not equal to), and `>=` (greater than or equal to). When these
comparisons are true (as in `(= 1 1)`), they return `T` and when they aren't (as
in `(> 0 1)`), they return `NIL`. For a full list of comparison operators, see

> [here](http://clhs.lisp.se/Body/f_eq_sle.htm).

Reviewer: The [docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd) state:

> Links should be used sparingly, if at all. While a link explaining a complex topic like recursion might be useful, for most concepts the links will provide more information than needed so explaining things concisely inline should be the aim.

Maybe remove the link to the comparison operators? It could be in the `hints.md` and/or `after.md` instead.
